File: ugly-number-ii.cpp
//Time - O(n)
//Space - O(1)
class Solution {
public:
    int nthUglyNumber(int n) {
        if(n==1) return 1;
        int a = 0,b = 0,c = 0,u;
        vector<int> k(1,1);
        
        while(k.size()<n){
            u = min(k[a]*2,min(k[b]*3,k[c]*5));
            k.push_back(u);
            if(k.back() == k[a]*2) a++;
            if(k.back() == k[b]*3) b++;
            if(k.back() == k[c]*5) c++;
            
        }
        
        return k.back();
    }
};

File: Ugly Number II.py
# Time Complexity :  O(n)
# Space Complexity : O(n)
# Did this code successfully run on Leetcode : yes
# Any problem you faced while coding this : no

class Solution:
    def nthUglyNumber(self, n: int) -> int:
        ugly = [1]
        p2 = p3 = p5 = 0
        for i in range(1,n):
            u2 = 2*ugly[p2]
            u3 = 3*ugly[p3]
            u5 = 5*ugly[p5]

            u = min(u2,u3,u5)
            ugly.append(u)

            if u == u2:
                p2 += 1
            if u == u3:
                p3 += 1
            if u == u5:
                p5 += 1
        return ugly[n-1]